bip-0442.md

```diff
@@ -90,138 +91,187 @@ highly optimized for just 2 stack elements.
+| :--------------- | --------: | -------: | ------: | ------: | ------: | :----: |
+| APO-Annex        |      8 WU |   113 WU |  100 WU | 1221 WU |  627 WU | SigOp  |
+| APO-Return       |      8 WU |   113 WU |   66 WU | 1359 WU |  765 WU | SigOp  |
+| CTV+CSFS         |     43 WU |    81 WU |   98 WU | 1394 WU |  765 WU |  CTV   |
+| CTV+CSFS+IKEY    |     10 WU |    48 WU |   98 WU | 1328 WU |  732 WU |  CTV   |
+| CTV+CSFS+IKEY+PC |     11 WU |    49 WU |  131 WU | 1191 WU |  594 WU |  CTV   |
 
 *ChannelSc: channel script, UpdateSc: update script, UpdateW: witness is the
-same size for both Force Close and Contest in LN-Symmetry, ForceC: total cost of unilateral close transactions*
+same size for both Force Close and Contest in Lightning Symmetry, ForceC: total
+cost of unilateral close transactions, Contest: The additional cost to contest
+a force closure, Settle: Whether a signature operation or CTV operation is
+required to validate the settlement transaction*
+
+### Proving general computation using trees
+
+One potential risk of PAIRCOMMIT is that it may enable verification of general
+computations on bitcoin.
 
-### Proving general computation
+Any script change which enables commitment to pairs of items necessarily
+enables commitments to Merkle trees.
 
 Merkle trees can be used to prove computation where the root of the tree
 represents the *function* and the leaves represent the *inputs* and *output*.
-There are practical limits to the entropy space for the *inputs* as they need
-to be iterated over and hashed into a Merkle root.
+There are practical limits to the total number of discrete output values that
+can be reached by all possible input values (i.e. the input entropy space) as
+they must be enumerated and hashed into a Merkle root.
 
-Taproot MAST trees can currently cover 128 bits of entropy space, which is over
-the practical limits to iterate over and merklize. Therefore, we conclude this
-capability does not materially extend what computations are possible to prove
-in bitcoin script. While `OP_PAIRCOMMIT` is not limited to a height of 128,
-that should not be practically feasible to utilize.
-
-There is a way to reduce the size of the witness for proving computation,
-by eliminating the Merkle path inclusion proofs, using `OP_CHECKSIGFROMSTACK`
-together with `OP_PAIRCOMMIT`. This method involves deleted key assumptions,
-most likely using MPC to create an enormous amount of signatures for the stack
-elements representing the *inputs* and the *output* of the *function*.
+Taproot trees can be 128 levels deep, therefore including up to 2^128 possible
+output values. This is over the practical limits to enumerate and merkelize.
+Therefore, we conclude that enabling Merkle commitments in script does not
+materially extend what computations are possible to validate in bitcoin
+script. While `OP_PAIRCOMMIT` is not explicitly limited to a height of 128, greater
+heights are not practical to Merkelize.
```

Comment:
Now you have found three different ways to spell it. ;)
The researcher’s name is Ralph Merkle, so either "merklized" or "Merklized" seems right.

```suggestion
heights are not practical to Merklize.
```

Reply:
We had a discussion about this a few years ago on the optech repo and IIRC Ralph Merkle himself replied in favor of "merklize"...

https://github.com/bitcoinops/bitcoinops.github.io/blob/master/STYLE.md#spelling

Reply:
ah, here it is: https://github.com/bitcoinops/bitcoinops.github.io/issues/171#issuecomment-529167731